oceanprotocol/market
# Update Number to Decimal

Fixes #624 

## Files

- [x] [Conversion.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/atoms/Price/Conversion.tsx) - can’t be updated since the lib we’re using `coingecko` accepts only type number as a parameter
- [x] [PriceUnit.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/atoms/Price/PriceUnit.tsx) - can’t be updated since the lib we’re using `coingecko` accepts only type number as a parameter
- [x] [Details.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/molecules/Wallet/Details.tsx) - can’t be updated since the lib we’re using `coingecko` accepts only type number as a parameter
- [x] [Consume.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Consume.tsx) - I don’t think we need to change it here. We use `Number` only for a boolean check
- [x] [components/organisms/AssetActions/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/index.tsx) - I don’t think we need to change it here. We use `Number` only for a boolean check
- [x] [FormComputeDataset.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Compute/FormComputeDataset.tsx) - I don’t think we need to change it here. We use it for UI display
- [x] [components/organisms/AssetActions/Compute/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Compute/index.tsx) - I don’t think we need to change it here since we’re using it for boolean check
- [x] [components/organisms/AssetActions/Pool/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Pool/index.tsx)
- [x] [components/organisms/AssetActions/Pool/utils.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Pool/utils.ts)
- [x] [FormAdd.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Pool/Add/FormAdd.tsx)
- [x] [components/organisms/AssetActions/Pool/Add/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Pool/Add/index.tsx)
- [x] [FormTrade.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Trade/FormTrade.tsx)
- [x] [components/organisms/AssetActions/Trade/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Trade/index.tsx)
- [x] [Output.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Trade/Output.tsx)
- [x] [Swap.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Trade/Swap.tsx)
- [x] [TradeInput.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetActions/Trade/TradeInput.tsx) - I don’t think we need to change it here
- [x] [Dynamic.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetContent/Pricing/FormPricing/Dynamic.tsx) - I don’t think we need to change it here; used for UI
- [x] [components/organisms/AssetContent/Pricing/FormPricing/index.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetContent/Pricing/FormPricing/index.tsx)
- [x] [Price.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/organisms/AssetContent/Pricing/FormPricing/Price.tsx)
- [x] [PoolShares.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/pages/History/PoolShares.tsx)

## Changes proposed in this PR:

- Switching from `Number` to `Decimal` where is possible and necessary.
- Update types of some variables to use `string` - having a unified approach to passing numbers across the different react elements

In some places like [File.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/atoms/File.tsx#L34) and [Time.tsx](https://github.com/oceanprotocol/market/blob/main/src/components/atoms/Time.tsx#L21), we don't need to switch to `Decimal`.

## How to test

In some cases switching from `Number` to `Decimal` in the file is not enough since we're using different structures. Follow the commits and what changes have been made.

- Testing with edge cases (where is possible)

Things to test/look for:
- Before each conversion to `Decimal` we have `isValidNumber()` check since any value different from `number`, `string`, or `Decimal` will break the conversion.
- Some calculations are more complex (a lot of things are happening) and the syntax when using `Decimal` is a bit different. When reviewing check for valid flow of mathematical operations (best way: check the changes in the commit - old/new)


## Notes

- Since we're using `Decimal` passing `undefined`, `null` and etc to the function `new Decimal()` breaks the code. Unfortunately, when fetching data from other places it's impossible to completely avoid this. That's why before each `Decimal` conversion there is a check with `isValidNumber`, in the case it's not a valid number we set a default value which is updated as soon as the conversion to `Decimal` can pass
- In some file where we use `Number` in boolean checks I left it as it is. I don't think we need to change it there since it doesn't make much difference.
